### PR TITLE
Read Metal3 config from a new CRD

### PR DESCRIPTION
**Pull request by sadasu**
Add functionality for MAO to read provisionings.metal3.io using a dynamic client and updating the baremetal config with config items needed to provision baremetal hosts.

### CODE REVIEW

File: install/0000_30_machine-api-operator_04_metal3provisioning.crd.yaml [hunk 1]
```diff
@@ -15,9 +15,9 @@ spec:
   validation:
     openAPIV3Schema:
       description: 'Provisioning contains configuration used by the Provisioning
```
Comment (enxebre) [inline install/0000_30_machine-api-operator_04_metal3provisioning.crd.yaml]: this should go into a different commit

File: pkg/operator/baremetal_config.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package operator
+
+import (
+	"github.com/golang/glog"
+
```
Comment (enxebre) [inline pkg/operator/baremetal_config.go]: extra spaces here and 8

File: pkg/operator/baremetal_test.go [hunk 3]
```diff
@@ -3,10 +3,37 @@ package operator
 import (
 	"testing"
 
+	"sigs.k8s.io/yaml"
+
```
Comment (enxebre) [inline pkg/operator/baremetal_test.go]: extraspace

File: pkg/operator/baremetal_test.go [hunk 4]
```diff
@@ -3,10 +3,37 @@ package operator
...
+	fakedynamic "k8s.io/client-go/dynamic/fake"
 	fakekube "k8s.io/client-go/kubernetes/fake"
 )
 
+var yamlContent = `
+apiVersion: metal3.io/v1alpha1
+kind: Provisioning
+metadata:
+  name: test
+  labels:
```
Comment (enxebre) [inline pkg/operator/baremetal_test.go]: please remove this label, it has no meaning

File: pkg/operator/baremetal_config.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+func getBaremetalProvisioningConfig(dc dynamic.Interface, configName string, usingBareMetal bool) BaremetalConfig {
+	if !usingBareMetal {
+		return BaremetalConfig{}
+	}
+
+	provisioningClient := dc.Resource(provisioningGVR)
+
+	provisioningConfig, err := provisioningClient.Get(configName, metav1.GetOptions{})
+
```
Comment (enxebre) [inline pkg/operator/baremetal_config.go]: remove extra space

File: pkg/operator/baremetal_config.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	provisioningClient := dc.Resource(provisioningGVR)
+	provisioningConfig, err := provisioningClient.Get(configName, metav1.GetOptions{})
+	if err != nil {
+		glog.V(1).Infof("Error getting provisioning config %q", configName)
+		return BaremetalConfig{}
+	}
+	provisioningSpec, found, err := unstructured.NestedMap(provisioningConfig.UnstructuredContent(), "spec")
+	if !found {
+		glog.Errorf("Nested Spec not found in Baremetal provisioning CR %q", configName)
```
Comment (enxebre) [inline pkg/operator/baremetal_config.go]: don't you want to error here for all the bad cases? I'd suggest you move this function to be called by https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L77 instead of the current caller you can handle errors as you wish without impacting other platforms.
Also how is this going to handle defaults?

File: vendor/modules.txt [hunk 7]
```diff
@@ -347,33 +347,34 @@ k8s.io/apimachinery/pkg/util/runtime
 k8s.io/apimachinery/pkg/watch
```
Comment (enxebre) [inline vendor/modules.txt]: please move vendoring into its own commit

File: pkg/operator/baremetal_config.go (hunk 5 above)
Comment (sadasu) [reply pkg/operator/baremetal_config.go]: Done.

File: pkg/operator/baremetal_test.go (hunk 4 above)
Comment (sadasu) [reply pkg/operator/baremetal_test.go]: Done.

File: pkg/operator/baremetal_test.go (hunk 3 above)
Comment (sadasu) [reply pkg/operator/baremetal_test.go]: Done.

File: pkg/operator/baremetal_config.go (hunk 2 above)
Comment (sadasu) [reply pkg/operator/baremetal_config.go]: Done.

File: install/0000_30_machine-api-operator_04_metal3provisioning.crd.yaml (hunk 1 above)
Comment (sadasu) [reply install/0000_30_machine-api-operator_04_metal3provisioning.crd.yaml]: Agreed.

File: vendor/modules.txt (hunk 7 above)
Comment (enxebre) [reply vendor/modules.txt]: also this is missing packages to be vendored /go/src/github.com/openshift/machine-api-operator/vendor/k8s.io/client-go/dynamic/fake, please include them via
go mod tidy, go mod vendor

File: pkg/operator/baremetal_config.go (hunk 6 above)
Comment (sadasu) [reply pkg/operator/baremetal_config.go]: The current implementation will have to support reading from the CR first and if that fails, fall back to reading from the ConfigMap to support backwards compatibility among other things. So, when we hit errors, we plan to just return an empty BaremetalConfig struct. Logic that is coming in a following PR would see that we were not able to get config from the CR and fallback to  using the configmap.
https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/baremetal_pod.go#L17.

Regarding calling getBaremetalProvisioningConfig() from sync.go (https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L77 ). We added BaremetalConfig to OperatorConfig and ideally before we try to deploy metal3, OperatorConfig should be fully populated.
We could potentially, leave BaremetalConfig uninitialized until we try to deploy metal3. I was keeping the flow similar to the other baremetal related config, the BaremetalControllers.

File: vendor/modules.txt (hunk 7 above)
Comment (sadasu) [reply vendor/modules.txt]: https://github.com/openshift/machine-api-operator/pull/472 has been created to handling vendoring for k8s.io/client-go/dynamic/fake. Will mark this PR as WIP until the vendoring merges.

File: pkg/operator/baremetal_config.go (hunk 6 above)
Comment (enxebre) [reply pkg/operator/baremetal_config.go]: I see, so if you fail to read the CRD, or there's a missing field in the CRD, then you fallback to the configMap and if anything fails you return the empty struct. What if there's a transient failure to read from the API you just set the empty struct as well?
In any case I'd still prefer this to be in https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L77 so this is completely isolated from other platforms and you can change how to handle those errors anytime.

File: vendor/modules.txt (hunk 7 above)
Comment (enxebre) [reply vendor/modules.txt]: please include any vendoring required by this PR in this PR

File: pkg/operator/baremetal_config.go (hunk 6 above)
Comment (sadasu) [reply pkg/operator/baremetal_config.go]: That is correct. Updating to accommodate comment.

Comment (sadasu) [reply pkg/operator/baremetal_config.go]: Refactored to remove BaremetalConfig from OperatorConfig and moved initialization of BaremetalConfig to within https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L77 .

File: pkg/operator/baremetal_config.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	dynamic "k8s.io/client-go/dynamic"
+)
+
+var provisioningGVR = schema.GroupVersionResource{Group: "metal3.io", Resource: "provisionings", Version: "v1alpha1"}
+
+func getBaremetalProvisioningConfig(dc dynamic.Interface, configName string) BaremetalConfig {
+	provisioningClient := dc.Resource(provisioningGVR)
+	provisioningConfig, err := provisioningClient.Get(configName, metav1.GetOptions{})
+	if err != nil {
+		glog.V(1).Infof("Error getting provisioning config %q", configName)
```
Comment (enxebre) [inline pkg/operator/baremetal_config.go]: I find it weird to return `BaremetalConfig{}` here, at a minimum this should log error/warning rather than info

File: pkg/operator/baremetal_config.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	if !found {
+		glog.Errorf("provisioningDHCPExternal not found in Baremetal provisioning CR %q", configName)
+		return BaremetalConfig{}
+	}
+	provisioningDHCPRange, found, err := unstructured.NestedString(provisioningSpec, "provisioningDHCPRange")
+	if !found {
+		glog.Errorf("provisioningDHCPRange not found in Baremetal provisioning CR %q", configName)
+		return BaremetalConfig{}
+	}
+	return BaremetalConfig{
```
Comment (enxebre) [inline pkg/operator/baremetal_config.go]: wouldn't `BaremetalProvisioningConfig` be more appropriate?

File: pkg/operator/baremetal_test.go [hunk 10]
```diff
@@ -71,3 +95,31 @@ func TestCreateMariadbPasswordSecret(t *testing.T) {
...
+		t.Errorf("failed to unmarshall input yaml content:%v", err)
+	}
+	dynamicClient := fakedynamic.NewSimpleDynamicClient(runtime.NewScheme(), u)
+	baremetalConfig := getBaremetalProvisioningConfig(dynamicClient, "test")
+	if baremetalConfig.ProvisioningInterface != expectedProvisioningInterface ||
+		baremetalConfig.ProvisioningIp != expectedProvisioningIP ||
+		baremetalConfig.ProvisioningNetworkCIDR != expectedProvisioningNetworkCIDR ||
+		baremetalConfig.ProvisioningDHCPExternal != expectedProvisioningDHCPExternal ||
+		baremetalConfig.ProvisioningDHCPRange != expectedProvisioningDHCPRange {
+		t.Errorf("failed getBaremetalProvisioningConfig. One or more BaremetalConfig items do not match the expected config.")
```
Comment (enxebre) [inline pkg/operator/baremetal_test.go]: can you render here the expected vs provisioned?

File: pkg/operator/config.go [hunk 11]
```diff
@@ -58,6 +58,15 @@ type Images struct {
 	BaremetalStaticIpManager     string `json:"baremetalStaticIpManager"`
 }
```
Comment (enxebre) [inline pkg/operator/config.go]: can we please declare this in `baremetal_config.go`

**Comment by enxebre** [comment]
@bison PTAL

File: pkg/operator/baremetal_config.go (hunk 8 above)
Comment (sadasu) [reply pkg/operator/baremetal_config.go]: Do you suggest a retry mechanism (good for transient errors) or a mechanism to bubble up an error (should not ignore errors) from this method?

File: pkg/operator/baremetal_pod.go [hunk 12]
```diff
@@ -98,7 +98,7 @@ func createMariadbPasswordSecret(client coreclientv1.SecretsGetter, config *Oper
 	return err
 }
 
-func newMetal3Deployment(config *OperatorConfig) *appsv1.Deployment {
+func newMetal3Deployment(config *OperatorConfig, baremetalConfig BaremetalConfig) *appsv1.Deployment {
```
Comment (bison) [inline pkg/operator/baremetal_pod.go]: I feel like I must be missing something. This reads the config from a new CRD so that it can pass it to this function, but then it appears to not be used.

Comment (sadasu) [reply pkg/operator/baremetal_pod.go]: @bison yes, that is coming in a following PR.

File: pkg/operator/config.go (hunk 11 above)
Comment (sadasu) [reply pkg/operator/config.go]: Done.

File: pkg/operator/baremetal_config.go (hunk 8 above)
Comment (sadasu) [reply pkg/operator/baremetal_config.go]: Done.

File: pkg/operator/baremetal_test.go (hunk 10 above)
Comment (sadasu) [reply pkg/operator/baremetal_test.go]: Done.

File: pkg/operator/baremetal_pod.go (hunk 12 above)
Comment (sadasu) [reply pkg/operator/baremetal_pod.go]: Using it is coming in a subsequent PR. Adding TODO to indicate that in code.

File: pkg/operator/baremetal_test.go [hunk 13]
```diff
@@ -71,3 +95,40 @@ func TestCreateMariadbPasswordSecret(t *testing.T) {
...
+func TestGetBaremetalProvisioningConfig(t *testing.T) {
+	u := &unstructured.Unstructured{Object: map[string]interface{}{}}
+	if err := yaml.Unmarshal([]byte(yamlContent), &u); err != nil {
+		t.Errorf("failed to unmarshall input yaml content:%v", err)
+	}
+	dynamicClient := fakedynamic.NewSimpleDynamicClient(runtime.NewScheme(), u)
+	baremetalConfig, err := getBaremetalProvisioningConfig(dynamicClient, "test")
+	if err != nil {
+		t.Logf("Unstructed Config:  %+v", u)
+		t.Fatalf("Failed to get Baremetal Provisioning Interface from CR %s", "test")
```
Comment (enxebre) [inline pkg/operator/baremetal_test.go]: no need to use %s here since you are passing a hardcoded string

File: pkg/operator/baremetal_test.go [hunk 14]
```diff
@@ -71,3 +95,40 @@ func TestCreateMariadbPasswordSecret(t *testing.T) {
...
+	}
+	if baremetalConfig.ProvisioningInterface != expectedProvisioningInterface ||
+		baremetalConfig.ProvisioningIp != expectedProvisioningIP ||
+		baremetalConfig.ProvisioningNetworkCIDR != expectedProvisioningNetworkCIDR ||
+		baremetalConfig.ProvisioningDHCPExternal != expectedProvisioningDHCPExternal ||
+		baremetalConfig.ProvisioningDHCPRange != expectedProvisioningDHCPRange {
+		t.Logf("Actual BaremetalProvisioningConfig: %+v", baremetalConfig)
+		t.Logf("Expected : ProvisioningInterface: %s, ProvisioningIP: %s, ProvisioningNetworkCIDR: %s, ProvisioningDHCPExternal: %t, expectedProvisioningDHCPRange: %s", expectedProvisioningInterface, expectedProvisioningIP, expectedProvisioningNetworkCIDR, expectedProvisioningDHCPExternal, expectedProvisioningDHCPRange)
```
Comment (enxebre) [inline pkg/operator/baremetal_test.go]: it'd be nice to follow `got: n%v, expected: %v` format as everywhere else

File: pkg/operator/sync.go [hunk 15]
```diff
@@ -74,13 +74,18 @@ func (optr *Operator) syncClusterAPIController(config *OperatorConfig) error {
 }
 
 func (optr *Operator) syncBaremetalControllers(config *OperatorConfig) error {
-	// First create a Secret needed for the Metal3 deployment
+	// Try to get baremetal provisioning config from a CR
+	baremetalProvisioningConfig, err := getBaremetalProvisioningConfig(optr.dynamicClient, baremetalProvisioningCR)
+	if err != nil {
+		glog.Infof("Unable to read Baremetal Provisioning config from CR %s.", baremetalProvisioningCR)
```
Comment (enxebre) [inline pkg/operator/sync.go]: I find it weird not returning err here, at minimum this should log.error, and make clear in the message the function will continue nevertheless

File: pkg/operator/baremetal_test.go [hunk 16]
```diff
@@ -71,3 +95,40 @@ func TestCreateMariadbPasswordSecret(t *testing.T) {
...
+}
+
+func TestGetIncorrectBaremetalProvisioningCR(t *testing.T) {
+	u := &unstructured.Unstructured{Object: map[string]interface{}{}}
+	if err := yaml.Unmarshal([]byte(yamlContent), &u); err != nil {
+		t.Errorf("failed to unmarshall input yaml content:%v", err)
+	}
+	dynamicClient := fakedynamic.NewSimpleDynamicClient(runtime.NewScheme(), u)
+	baremetalConfig, err := getBaremetalProvisioningConfig(dynamicClient, "test1")
+	if err != nil {
```
Comment (enxebre) [inline pkg/operator/baremetal_test.go]: wouldn't be the client expected to error here and so `getBaremetalProvisioningConfig`

For units we prefer to have one test per function i.e `TestGetBaremetalProvisioningConfig` and loop over multiple test cases within it. e.g https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/vsphere/machine_scope_test.go#L139-L261

**Comment by enxebre** [comment]
nit: we use the imperative mood in the subject line for commits rather than -ing, https://chris.beams.io/posts/git-commit/

File: pkg/operator/baremetal_config.go (hunk 9 above)
Comment (sadasu) [reply pkg/operator/baremetal_config.go]: Done.

File: pkg/operator/sync.go (hunk 15 above)
Comment (sadasu) [reply pkg/operator/sync.go]: In this release, we cannot return error and stop deploying Metal3 because during discussions around upgrade strategy, we had decided to have one release where the config could be obtained from either sources, the ConfigMap or the newly created Config resource.
In the current implementation, we are prioritizing reading config from the new Config Resource. When that fails, (could be for a simple reason that the user did not want to provide a CR and instead still wanted to use a configmap), then we should be able to proceed with the deployment. 
The other part of the comment talks about logging an error. Again, I worry that this might be interpreted as something went wrong. If you think seeing an error msg in the logs will not be interpreted as I think it will, I am fine with making this Info statement into an Error msg.
